ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java

```diff
@@ -1,18 +1,93 @@
 package com.microsoft.azure.kusto.ingest.resources;
 
+import com.microsoft.azure.kusto.ingest.utils.TimeProvider;
+
+import java.util.ArrayDeque;
+
 public class RankedStorageAccount {
+    static class Bucket {
+        public int successCount;
+        public int failureCount;
+
+        public Bucket() {
+            this.successCount = 0;
+            this.failureCount = 0;
+        }
+    }
+
+    private final ArrayDeque<Bucket> buckets = new ArrayDeque<>();
     private final String accountName;
+    private final int maxNumberOfBuckets;
+    private final int bucketDurationInSec;
+    private final TimeProvider timeProvider;
+
+    private long lastActionTimestamp;
 
-    public RankedStorageAccount(String accountName) {
+    public RankedStorageAccount(String accountName, int maxNumberOfBuckets, int bucketDurationInSec, TimeProvider timeProvider) {
         this.accountName = accountName;
+        this.maxNumberOfBuckets = maxNumberOfBuckets;
+        this.bucketDurationInSec = bucketDurationInSec;
+        this.timeProvider = timeProvider;
+        lastActionTimestamp = timeProvider.currentTimeMillis();
     }
 
     public void addResult(boolean success) {
+        Bucket lastBucket = adjustForTimePassed();
+        if (success) {
+            lastBucket.successCount++;
+        } else {
+            lastBucket.failureCount++;
+        }
+    }
 
+    private Bucket adjustForTimePassed() {
+        if (buckets.isEmpty()) {
+            buckets.push(new Bucket());
+        }
+
+        long timePassed = timeProvider.currentTimeMillis() - lastActionTimestamp;
+        long bucketsToCreate = timePassed / (bucketDurationInSec * 1000L);
+        if (bucketsToCreate >= maxNumberOfBuckets) {
+            buckets.clear();
+            buckets.push(new Bucket());
```

Comment:
Why is this needed? Not sure I understand.
Wouldn't it be polled anyway in line 58?

Reply:
Good catch - the idea is to save computation time when we need to throw all the buckets away, there should have been an else in there, I'll fix.